### DISCUSSION TITLE
Halo 2 是怎么压缩多步运算的证明的呢？

### DISCUSSION THREAD

**Discussion by voidcenter** [z2o-k7e/halo2-step-by-step#15] [Q&A]
我有一个 high-level 的问题，如果一个程序很长，有很多步，就需要对每一步产生证明然后把证明压缩起来
按我的理解，两种压缩方法分别是递归和折叠，
halo 2是用哪一种方法啊？

**Comment by Demian101** [comment]
<img width="782" alt="image" src="https://github.com/zkp-co-learning/halo2-step-by-step/assets/33189338/91b3b4e2-ac23-4c1e-b3b2-45ff93d9dd45">

“每一步产生证明然后把证明压缩起来”  这里我没理解，为什么要对每一步产生证明 🤣

**Reply by mdj33** [reply]
我理解这里的很多步的程序，相当于有很多门，或者很多个子程序，证明压缩起来的方法，一般是把产生的proof聚合成一个proof
如果是nova机制，就是折叠，在产生proof之前把程序折叠起来，然后再产生proof
halo2都可以支持，目前是支持的聚合的方法，这是证明的后端，为了支持折叠，可以对halo2改造，相当于改造证明的前端

**Reply by voidcenter** [reply]
我讲的不太清楚，我是想说有时候计算大过电路的上限了就要切成很多小块，每个小块证明，然后再用递归证明合起来。还有一种情况要切割的应该是在多方计算的时候，每方要对自己的步骤产生一个证明，然后大家的证明合起来。

“目前是支持的聚合的方法”，聚合的方法是什么啊？
